### PR TITLE
chore(expo-asset): remove unused web features

### LINKED ISSUE #131: [WEB] Attempting to load assets on nested route fails
### Summary

Loading assets from a 'base' route (e.g. `/somepath`) works correctly, loading them from `/assets/[path to assets]`.

When attempting to load assets such as fonts from a nested path (e.g. `/nested/123`), the page attempts to load them from `/nested/assets/[path to assets]` which is incorrect and leads to the page not rendering correctly. This only happens if you visit the page URL directly, or refresh the browser while on that page.

This happens with locally-sourced assets (e.g. those hosted inside the `assets` folder) but also with node_modules (e.g. the `expo-google-fonts` packages exhibit the same behaviour).

This test was done on the **web** platform - I haven't explicitly tested on mobile/native as there is no way to refresh the page to trigger the bug.

For clarification, I'm unsure if this is a router-specific issue or a bundler issue. Please move to the appropriate repo if this is not the correct one.

### Minimal reproducible example

https://github.com/ChronSyn/expo-router_bug-report-assetpath

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- fix https://github.com/expo/router/issues/131 by using the correct web base URL, previously was failing to resolve the origin using a number of naive checks.
- drop all incorrect offline checks for web assets
- utilize the newer @react-native/assets/registry package
- reuse upstream types
- fix file system type error
- Possibly breaks web snack support introduced here https://github.com/expo/expo/pull/14435
- ~~Also might break some assumptions about Expo Webpack, will need to do some testing.~~ Webpack just ignores all of this code.

# How

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-asset/src/AssetSourceResolver.web.ts [hunk 1]
```diff
@@ -52,54 +31,47 @@ export default class AssetSourceResolver {
...
-    const path = this.jsbundleUrl || '';
-    return this.fromSource(path + getScaledAssetPath(this.asset));
-  }
-  resourceIdentifierWithoutScale(): ResolvedAssetSource {
-    throw new UnavailabilityError('react-native', 'resourceIdentifierWithoutScale()');
-  }
-  drawableFolderInBundle(): ResolvedAssetSource {
-    throw new UnavailabilityError('react-native', 'drawableFolderInBundle()');
+    const fromUrl = new URL(this.serverUrl);
+    fromUrl.pathname = getScaledAssetPath(this.asset);
```
Comment (ide) [inline packages/expo-asset/src/AssetSourceResolver.web.ts:62]: Should this preserve the pathname in the serverUrl? For instance, if serverUrl is "https://example.com/hello", then should an asset at "icon.png" or "./icon.png" resolve to "https://example.com/hello/icon.png" or "https://example.com/icon.png"?

```suggestion
    // conceptually speaking (path.join() is a Node function)
    fromUrl.pathname = path.join(fromUrl.pathname, getScaledAssetPath(this.asset));
```

The main use case I can think of for this is using a cloud storage bucket, which often has a base URL like "region.cloudproviderstorage.com/bucket".

Or do we expect paths in the base URL to be handled by `asset.httpServerLocation`?

File: packages/expo-asset/src/__tests__/Asset-test.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ jest.mock('../ImageAssets', () => {
   const ImageAssets = jest.requireActual('../ImageAssets');
   return {
     ...ImageAssets,
-    getImageInfoAsync: jest.fn(),
+    getImageInfoAsync: jest.fn(() => Promise.resolve({ width: 1, height: 1 })),
```
Comment (ide) [inline packages/expo-asset/src/__tests__/Asset-test.ts:30]: Doesn't matter much, just pointing out that I like to mock async functions with async functions themselves:

```suggestion
    getImageInfoAsync: jest.fn(async () => ({ width: 1, height: 1 })),
```

**Review by ide — COMMENTED** [review]
Looks good mostly. Just one main discussion point about pathnames in an inline comment.

**Review by ide — APPROVED** [review]

**Comment by ide** [comment]
It looks like a lot of the changes from earlier revisions of this PR are now gone. Was that intentional?

**Comment by EvanBacon** [comment]
@ide good catch, not sure what happened there. Fixed, and manually pulled into an Expo Router project as a sanity test.

**Comment by codinsonn** [comment]
@EvanBacon Any pointers on what to do when you get errors about missing assets? e.g.
```
Module "12" is missing from the asset registry
```
Worked fine before sdk 48

**Pull request merged**